langchain/src/tools/outlook/authFlowREST.ts

```diff
@@ -0,0 +1,205 @@
+import * as http from "http";
+import * as url from "url";
+import { AuthFlowBase } from "./authFlowBase.js";
+import { getEnvironmentVariable } from "../../util/env.js";
+
+interface AccessTokenResponse {
+  access_token: string;
+  refresh_token: string;
+}
+
+// this class uses your clientId, clientSecret and redirectUri
+// to get access token and refresh token, if you already have them,
+// you can use AuthFlowToken or AuthFlowRefresh instead
+export class AuthFlowREST extends AuthFlowBase {
+  private clientSecret: string;
+
+  private redirectUri: string;
+
+  private port: number;
+
+  private pathname: string;
+
+  private refreshToken = "";
+
+  constructor({
+    clientId,
+    clientSecret,
+    redirectUri,
+  }: { clientId?: string; clientSecret?: string; redirectUri?: string } = {}) {
+    let id = clientId;
+    let secret = clientSecret;
+    let uri = redirectUri;
+    if (!id || !secret || !uri) {
+      id = getEnvironmentVariable("OUTLOOK_CLIENT_ID");
+      secret = getEnvironmentVariable("OUTLOOK_CLIENT_SECRET");
+      uri = getEnvironmentVariable("OUTLOOK_REDIRECT_URI");
+    }
```

Comment:
We shouldn't override all if one isn't passed, instead lets do something like this:
```suggestion
    let id = clientId;
    let secret = clientSecret;
    let uri = redirectUri;
    if (!id || !secret || !uri) {
      id = id ?? getEnvironmentVariable("OUTLOOK_CLIENT_ID");
      secret = secret ??  getEnvironmentVariable("OUTLOOK_CLIENT_SECRET");
      uri = uri ?? getEnvironmentVariable("OUTLOOK_REDIRECT_URI");
    }
```